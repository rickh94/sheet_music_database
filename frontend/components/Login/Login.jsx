import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { withAlert } from 'react-alert'

import Container from 'react-bulma-components/lib/components/container'
import {
  Field,
  Control,
  Checkbox,
} from 'react-bulma-components/lib/components/form'
import Button from 'react-bulma-components/lib/components/button'
import Card from 'react-bulma-components/lib/components/card'
import Content from 'react-bulma-components/lib/components/card/components/content'
import Notification from 'react-bulma-components/lib/components/notification'

import Header from '../Header'
import TextFieldWithErrors from '../TextFieldWithErrors'
import { account } from '../../actions'
import alertText from '../../middleware/alertText'

export class LoginWrapper extends Component {
  static propTypes = {
    login: PropTypes.func.isRequired,
    alert: PropTypes.object.isRequired,
    account: PropTypes.object.isRequired
  }

  componentDidMount() {
    if (this.props.account.token) {
      this.props.alert.show(alertText('Already Logged In'))
      this.props.history.goBack()
    }
  }

  attemptLogin = async (email, password, remember) => {
    let success
    try {
      success = await this.props.login(email, password, remember)
    } catch (err) {
      success = false
    }
    if (success) {
      this.props.alert.show(alertText('Login Successful'), { type: 'success' })
      this.props.history.goBack()
    } else {
      this.props.alert.show(alertText('Login Failed'), { type: 'error' })
      return this.props.account.errors
    }
  }

  render() {
    return (
      // Perhaps refactor to be centered fixed width
      <React.Fragment>
        <Header />
        <Container>
          <Card className="margin-default">
            <Card.Header>
              <Card.Header.Title>Login</Card.Header.Title>
            </Card.Header>
            <Card.Content>
              <LoginForm attemptLogin={this.attemptLogin} cancel={() => this.props.history.goBack()} />
              <Content>
                <Link to="/register">Don{"'"}t have an account? Register Here</Link>
              </Content>
            </Card.Content>
          </Card>
        </Container>
      </React.Fragment>
    )
  }
}

const mapStateToProps = state => {
  return {
    account: state.account
  }
}

const mapDispatchToProps = dispatch => {
  return {
    login: (email, password, remember) => {
      return dispatch(account.login(email, password, remember))
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withAlert(LoginWrapper))

export class LoginForm extends Component {
  state = {
    email: '',
    password: '',
    remember: false,
    errors: {
      email: null,
      password: null,
      nonFieldErrors: null
    }
  }

  static propTypes = {
    attemptLogin: PropTypes.func.isRequired,
    cancel: PropTypes.func.isRequired
  }

  onCancelClicked = () => {
    this.props.cancel()
  }

  onFieldChange = (e, name) => {
    this.setState({ [name]: e.target.value })
  }

  onLoginClicked = async e => {
    e.preventDefault()
    const { email, password, remember } = this.state
    if (!email) {
      this.setState({ errors: { email: 'Email is required' } })
      return
    }
    if (!password) {
      this.setState({ errors: { password: 'Password is required' } })
      return
    }
    const loginErrors = await this.props.attemptLogin(email, password, remember)
    if (loginErrors) {
      this.setState({
        errors: {
          email: loginErrors.email,
          password: loginErrors.password,
          nonFieldErrors: loginErrors.non_field_errors
        }
      })
    }
  }

  render() {
    const { email, password, remember, errors } = this.state
    return (
      <form>
        {errors.nonFieldErrors && (
          <Notification color="danger">
            {errors.nonFieldErrors}
            <Button
              remove
              onClick={() => this.setState({ errors: { nonFieldErrors: null } })}
            />
          </Notification>
        )}
        <TextFieldWithErrors
          label="Email"
          placeholder="Enter Email..."
          value={email}
          type="email"
          onChange={this.onFieldChange}
          error={errors.email}
          name="email"
        />
        <TextFieldWithErrors
          label="Password"
          placeholder="Enter Password..."
          value={password}
          type="password"
          error={errors.password}
          name="password"
          onChange={this.onFieldChange}
        />
        <Field>
          <Control>
            <Checkbox
              onChange={() => this.setState({ remember: !remember })}
              checked={remember}
            >
              {' '}
              Keep Me Logged In
            </Checkbox>
          </Control>
        </Field>
        <Field kind="group">
          <Control>
            <Button type="primary" onClick={e => this.onLoginClicked(e)}>
              Login
            </Button>
          </Control>
          <Control>
            <Button onClick={() => this.onCancelClicked()}>Cancel</Button>
          </Control>
        </Field>
      </form>
    )
  }
}
