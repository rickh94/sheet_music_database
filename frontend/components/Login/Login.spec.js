import { shallow } from 'enzyme'
import React from 'react'
import '../../setupTests'
import Login, { LoginForm } from './Login'

import { Label } from 'react-bulma-components/lib/components/form'

describe('Login', () => {
  const wrapper = shallow(<Login />)

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a container', () => {
    expect(wrapper.exists('Container')).toBeTruthy()
  })

  it('renders the Header', () => {
    expect(wrapper.exists('Header')).toBeTruthy()
  })

  it('renders a LoginForm', () => {
    expect(wrapper.exists('LoginForm')).toBeTruthy()
  })
})

describe('LoginForm', () => {
  const wrapper = shallow(<LoginForm />)

  beforeEach(() => {
    wrapper.update()
  })

  it('renders a form', () => {
    expect(wrapper.exists('form')).toBeTruthy()
  })

  it('renders needed fields', () => {
    expect(wrapper.find('Field').length).toEqual(3)
  })

  it('renders labels for email, password', () => {
    expect(wrapper.containsMatchingElement(<Label>Email</Label>)).toBeTruthy()
    expect(wrapper.containsMatchingElement(<Label>Password</Label>)).toBeTruthy()
  })

  it('renders fields with controls', () => {
    wrapper.find('Field').forEach(field => {
      expect(field.exists('Control')).toBeTruthy()
    })
  })

  it('renders email input for email field', () => {
    const field = wrapper.findWhere(n => n.contains('Email') && n.name() == 'Field')
    expect(field.exists('Input')).toBeTruthy()
    const input = field.find('Input')
    expect(input.props().type).toEqual('email')
  })

  it('renders password input for password field', () => {
    const field = wrapper.findWhere(el => el.contains('Password') && el.name() == 'Field')
    expect(field.exists('Input')).toBeTruthy()
    const input = field.find('Input')
    expect(input.props().type).toEqual('password')
  })

  it('renders a login button', () => {
    const loginButton = wrapper.findWhere(el => el.contains('Login') && el.name() == 'Button')
    expect(loginButton.length).toEqual(1)
  })

  it('calls login function when button is clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'onLoginClicked')
    wrapper.findWhere(el => el.contains('Login') && el.name() == 'Button').simulate('click')
    expect(spy).toHaveBeenCalled()
  })

  it('renders a cancel button', () => {
    const cancelButton = wrapper.findWhere(el => el.contains('Cancel') && el.name() == 'Button')
    expect(cancelButton.length).toEqual(1)
  })

  it('calls cancel function when button is clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'onCancelClicked')
    wrapper.findWhere(el => el.contains('Cancel') && el.name() == 'Button').simulate('click')
    expect(spy).toHaveBeenCalled()
  })
})
