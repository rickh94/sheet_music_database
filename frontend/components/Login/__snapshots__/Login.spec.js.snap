// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginWrapper renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginWrapper
    account={
      Object {
        "errors": Object {},
      }
    }
    alert={
      Object {
        "show": [MockFunction],
      }
    }
    history={
      Object {
        "goBack": [MockFunction],
      }
    }
    login={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(Header) />,
        <Container
          breakpoint={null}
          fluid={false}
          renderAs="div"
          style={Object {}}
        >
          <Card
            className="margin-default"
            renderAs="div"
            style={Object {}}
          >
            <CardHeader
              renderAs="div"
            >
              <CardHeaderTitle
                renderAs="div"
              >
                Login
              </CardHeaderTitle>
            </CardHeader>
            <CardContent
              renderAs="div"
            >
              <LoginForm
                attemptLogin={[Function]}
                cancel={[Function]}
              />
              <CardContent
                renderAs="div"
              >
                <Link
                  replace={false}
                  to="/register"
                >
                  Don
                  '
                  t have an account? Register Here
                </Link>
              </CardContent>
            </CardContent>
          </Card>
        </Container>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "backgroundColor": undefined,
          "breakpoint": null,
          "children": <Card
            className="margin-default"
            renderAs="div"
            style={Object {}}
          >
            <CardHeader
              renderAs="div"
            >
              <CardHeaderTitle
                renderAs="div"
              >
                Login
              </CardHeaderTitle>
            </CardHeader>
            <CardContent
              renderAs="div"
            >
              <LoginForm
                attemptLogin={[Function]}
                cancel={[Function]}
              />
              <CardContent
                renderAs="div"
              >
                <Link
                  replace={false}
                  to="/register"
                >
                  Don
                  '
                  t have an account? Register Here
                </Link>
              </CardContent>
            </CardContent>
          </Card>,
          "className": undefined,
          "clearfix": undefined,
          "clipped": undefined,
          "fluid": false,
          "hidden": undefined,
          "invisible": undefined,
          "italic": undefined,
          "marginless": undefined,
          "overlay": undefined,
          "paddingless": undefined,
          "pull": undefined,
          "radiusless": undefined,
          "renderAs": "div",
          "responsive": undefined,
          "shadowless": undefined,
          "style": Object {},
          "textAlignment": undefined,
          "textColor": undefined,
          "textSize": undefined,
          "textTransform": undefined,
          "textWeight": undefined,
          "unselectable": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "backgroundColor": undefined,
            "children": Array [
              <CardHeader
                renderAs="div"
              >
                <CardHeaderTitle
                  renderAs="div"
                >
                  Login
                </CardHeaderTitle>
              </CardHeader>,
              <CardContent
                renderAs="div"
              >
                <LoginForm
                  attemptLogin={[Function]}
                  cancel={[Function]}
                />
                <CardContent
                  renderAs="div"
                >
                  <Link
                    replace={false}
                    to="/register"
                  >
                    Don
                    '
                    t have an account? Register Here
                  </Link>
                </CardContent>
              </CardContent>,
            ],
            "className": "margin-default",
            "clearfix": undefined,
            "clipped": undefined,
            "hidden": undefined,
            "invisible": undefined,
            "italic": undefined,
            "marginless": undefined,
            "overlay": undefined,
            "paddingless": undefined,
            "pull": undefined,
            "radiusless": undefined,
            "renderAs": "div",
            "responsive": undefined,
            "shadowless": undefined,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "unselectable": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "backgroundColor": undefined,
                "children": <CardHeaderTitle
                  renderAs="div"
                >
                  Login
                </CardHeaderTitle>,
                "className": undefined,
                "clearfix": undefined,
                "clipped": undefined,
                "hidden": undefined,
                "invisible": undefined,
                "italic": undefined,
                "marginless": undefined,
                "overlay": undefined,
                "paddingless": undefined,
                "pull": undefined,
                "radiusless": undefined,
                "renderAs": "div",
                "responsive": undefined,
                "shadowless": undefined,
                "textAlignment": undefined,
                "textColor": undefined,
                "textSize": undefined,
                "textTransform": undefined,
                "textWeight": undefined,
                "unselectable": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "backgroundColor": undefined,
                  "children": "Login",
                  "className": undefined,
                  "clearfix": undefined,
                  "clipped": undefined,
                  "hidden": undefined,
                  "invisible": undefined,
                  "italic": undefined,
                  "marginless": undefined,
                  "overlay": undefined,
                  "paddingless": undefined,
                  "pull": undefined,
                  "radiusless": undefined,
                  "renderAs": "div",
                  "responsive": undefined,
                  "shadowless": undefined,
                  "textAlignment": undefined,
                  "textColor": undefined,
                  "textSize": undefined,
                  "textTransform": undefined,
                  "textWeight": undefined,
                  "unselectable": undefined,
                },
                "ref": null,
                "rendered": "Login",
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "backgroundColor": undefined,
                "children": Array [
                  <LoginForm
                    attemptLogin={[Function]}
                    cancel={[Function]}
                  />,
                  <CardContent
                    renderAs="div"
                  >
                    <Link
                      replace={false}
                      to="/register"
                    >
                      Don
                      '
                      t have an account? Register Here
                    </Link>
                  </CardContent>,
                ],
                "className": undefined,
                "clearfix": undefined,
                "clipped": undefined,
                "hidden": undefined,
                "invisible": undefined,
                "italic": undefined,
                "marginless": undefined,
                "overlay": undefined,
                "paddingless": undefined,
                "pull": undefined,
                "radiusless": undefined,
                "renderAs": "div",
                "responsive": undefined,
                "shadowless": undefined,
                "textAlignment": undefined,
                "textColor": undefined,
                "textSize": undefined,
                "textTransform": undefined,
                "textWeight": undefined,
                "unselectable": undefined,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "attemptLogin": [Function],
                    "cancel": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "backgroundColor": undefined,
                    "children": <Link
                      replace={false}
                      to="/register"
                    >
                      Don
                      '
                      t have an account? Register Here
                    </Link>,
                    "className": undefined,
                    "clearfix": undefined,
                    "clipped": undefined,
                    "hidden": undefined,
                    "invisible": undefined,
                    "italic": undefined,
                    "marginless": undefined,
                    "overlay": undefined,
                    "paddingless": undefined,
                    "pull": undefined,
                    "radiusless": undefined,
                    "renderAs": "div",
                    "responsive": undefined,
                    "shadowless": undefined,
                    "textAlignment": undefined,
                    "textColor": undefined,
                    "textSize": undefined,
                    "textTransform": undefined,
                    "textWeight": undefined,
                    "unselectable": undefined,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": Array [
                        "Don",
                        "'",
                        "t have an account? Register Here",
                      ],
                      "replace": false,
                      "to": "/register",
                    },
                    "ref": null,
                    "rendered": Array [
                      "Don",
                      "'",
                      "t have an account? Register Here",
                    ],
                    "type": [Function],
                  },
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(Header) />,
          <Container
            breakpoint={null}
            fluid={false}
            renderAs="div"
            style={Object {}}
          >
            <Card
              className="margin-default"
              renderAs="div"
              style={Object {}}
            >
              <CardHeader
                renderAs="div"
              >
                <CardHeaderTitle
                  renderAs="div"
                >
                  Login
                </CardHeaderTitle>
              </CardHeader>
              <CardContent
                renderAs="div"
              >
                <LoginForm
                  attemptLogin={[Function]}
                  cancel={[Function]}
                />
                <CardContent
                  renderAs="div"
                >
                  <Link
                    replace={false}
                    to="/register"
                  >
                    Don
                    '
                    t have an account? Register Here
                  </Link>
                </CardContent>
              </CardContent>
            </Card>
          </Container>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "backgroundColor": undefined,
            "breakpoint": null,
            "children": <Card
              className="margin-default"
              renderAs="div"
              style={Object {}}
            >
              <CardHeader
                renderAs="div"
              >
                <CardHeaderTitle
                  renderAs="div"
                >
                  Login
                </CardHeaderTitle>
              </CardHeader>
              <CardContent
                renderAs="div"
              >
                <LoginForm
                  attemptLogin={[Function]}
                  cancel={[Function]}
                />
                <CardContent
                  renderAs="div"
                >
                  <Link
                    replace={false}
                    to="/register"
                  >
                    Don
                    '
                    t have an account? Register Here
                  </Link>
                </CardContent>
              </CardContent>
            </Card>,
            "className": undefined,
            "clearfix": undefined,
            "clipped": undefined,
            "fluid": false,
            "hidden": undefined,
            "invisible": undefined,
            "italic": undefined,
            "marginless": undefined,
            "overlay": undefined,
            "paddingless": undefined,
            "pull": undefined,
            "radiusless": undefined,
            "renderAs": "div",
            "responsive": undefined,
            "shadowless": undefined,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "unselectable": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "backgroundColor": undefined,
              "children": Array [
                <CardHeader
                  renderAs="div"
                >
                  <CardHeaderTitle
                    renderAs="div"
                  >
                    Login
                  </CardHeaderTitle>
                </CardHeader>,
                <CardContent
                  renderAs="div"
                >
                  <LoginForm
                    attemptLogin={[Function]}
                    cancel={[Function]}
                  />
                  <CardContent
                    renderAs="div"
                  >
                    <Link
                      replace={false}
                      to="/register"
                    >
                      Don
                      '
                      t have an account? Register Here
                    </Link>
                  </CardContent>
                </CardContent>,
              ],
              "className": "margin-default",
              "clearfix": undefined,
              "clipped": undefined,
              "hidden": undefined,
              "invisible": undefined,
              "italic": undefined,
              "marginless": undefined,
              "overlay": undefined,
              "paddingless": undefined,
              "pull": undefined,
              "radiusless": undefined,
              "renderAs": "div",
              "responsive": undefined,
              "shadowless": undefined,
              "style": Object {},
              "textAlignment": undefined,
              "textColor": undefined,
              "textSize": undefined,
              "textTransform": undefined,
              "textWeight": undefined,
              "unselectable": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "backgroundColor": undefined,
                  "children": <CardHeaderTitle
                    renderAs="div"
                  >
                    Login
                  </CardHeaderTitle>,
                  "className": undefined,
                  "clearfix": undefined,
                  "clipped": undefined,
                  "hidden": undefined,
                  "invisible": undefined,
                  "italic": undefined,
                  "marginless": undefined,
                  "overlay": undefined,
                  "paddingless": undefined,
                  "pull": undefined,
                  "radiusless": undefined,
                  "renderAs": "div",
                  "responsive": undefined,
                  "shadowless": undefined,
                  "textAlignment": undefined,
                  "textColor": undefined,
                  "textSize": undefined,
                  "textTransform": undefined,
                  "textWeight": undefined,
                  "unselectable": undefined,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "backgroundColor": undefined,
                    "children": "Login",
                    "className": undefined,
                    "clearfix": undefined,
                    "clipped": undefined,
                    "hidden": undefined,
                    "invisible": undefined,
                    "italic": undefined,
                    "marginless": undefined,
                    "overlay": undefined,
                    "paddingless": undefined,
                    "pull": undefined,
                    "radiusless": undefined,
                    "renderAs": "div",
                    "responsive": undefined,
                    "shadowless": undefined,
                    "textAlignment": undefined,
                    "textColor": undefined,
                    "textSize": undefined,
                    "textTransform": undefined,
                    "textWeight": undefined,
                    "unselectable": undefined,
                  },
                  "ref": null,
                  "rendered": "Login",
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "backgroundColor": undefined,
                  "children": Array [
                    <LoginForm
                      attemptLogin={[Function]}
                      cancel={[Function]}
                    />,
                    <CardContent
                      renderAs="div"
                    >
                      <Link
                        replace={false}
                        to="/register"
                      >
                        Don
                        '
                        t have an account? Register Here
                      </Link>
                    </CardContent>,
                  ],
                  "className": undefined,
                  "clearfix": undefined,
                  "clipped": undefined,
                  "hidden": undefined,
                  "invisible": undefined,
                  "italic": undefined,
                  "marginless": undefined,
                  "overlay": undefined,
                  "paddingless": undefined,
                  "pull": undefined,
                  "radiusless": undefined,
                  "renderAs": "div",
                  "responsive": undefined,
                  "shadowless": undefined,
                  "textAlignment": undefined,
                  "textColor": undefined,
                  "textSize": undefined,
                  "textTransform": undefined,
                  "textWeight": undefined,
                  "unselectable": undefined,
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "attemptLogin": [Function],
                      "cancel": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "backgroundColor": undefined,
                      "children": <Link
                        replace={false}
                        to="/register"
                      >
                        Don
                        '
                        t have an account? Register Here
                      </Link>,
                      "className": undefined,
                      "clearfix": undefined,
                      "clipped": undefined,
                      "hidden": undefined,
                      "invisible": undefined,
                      "italic": undefined,
                      "marginless": undefined,
                      "overlay": undefined,
                      "paddingless": undefined,
                      "pull": undefined,
                      "radiusless": undefined,
                      "renderAs": "div",
                      "responsive": undefined,
                      "shadowless": undefined,
                      "textAlignment": undefined,
                      "textColor": undefined,
                      "textSize": undefined,
                      "textTransform": undefined,
                      "textWeight": undefined,
                      "unselectable": undefined,
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "children": Array [
                          "Don",
                          "'",
                          "t have an account? Register Here",
                        ],
                        "replace": false,
                        "to": "/register",
                      },
                      "ref": null,
                      "rendered": Array [
                        "Don",
                        "'",
                        "t have an account? Register Here",
                      ],
                      "type": [Function],
                    },
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
